dash/dash-renderer/src/actions/callbacks.ts

```diff
@@ -502,32 +504,85 @@ export function executeCallback(
 
                 if (clientside_function) {
                     try {
-                        resolve({
+                        return {
                             data: handleClientside(
                                 dispatch,
                                 clientside_function,
                                 config,
                                 payload
                             ),
                             payload
-                        });
+                        };
                     } catch (error) {
-                        resolve({error, payload});
+                        return {error, payload};
                     }
-                    return null;
                 }
 
-                handleServerside(dispatch, hooks, config, payload)
-                    .then(data => resolve({data, payload}))
-                    .catch(error => resolve({error, payload}));
+                let newConfig = config;
+                let newHeaders: Record<string, string> | null = null;
+                let retry = 0;
+
+                while (true) {
```

Comment:
`while` loops always make me uneasy... can this one also be a `for` like in `apiThunk`?

Reply:
✅ Changed to 'for' now. A tad more awkward to get typescript to respect the return type but I agree overall this is better now.